src/fs/service.go

```diff
@@ -650,3 +677,60 @@ func (service *MercuryFsService) uploadFile(writer http.ResponseWriter, request
 
 	return
 }
+
+// rename the given file path
+func renameFile(p string) string {
+	//get the file suffix
+	ext := path.Ext(p)
+	//get the filename without suffix
+	baseName := strings.Replace(p, ext, "", 1)
+
+	timeStamp := time.Now().Format("20060102-1504")
+
+	return baseName + "-" + timeStamp + ext
+}
+
+/*
+status:
+0 => file not exists,
+1 => file exists,
+2 => file exists and have the same md5
+*/
+func checkFileExists(filename, md5 string) int {
+	_, err := os.Stat(filename)
+	//file not exists
+	if err != nil {
+		return 0
+	}
+	file, ferr := os.Open(filename)
+	if ferr != nil {
+		return 0
+	}
+
+	//file exists, compare the md5
+	if CalMD5(file) == md5 {
+		return 2
+	}
+
+	//file exists
+	return 1
+}
+
+//check if the file name is valid
+func validFilename(f string) bool {
+	if strings.HasPrefix(f, ".") ||
+		strings.HasPrefix(f, "/") ||
+		//It could be slash after the system escaped it automatically.
+		strings.HasPrefix(f, "⁄") ||
+		strings.HasPrefix(f, " ") {
+		return false
```

Comment:
Only these many checks are not sufficient, there may be system reserved keywords and other things like max filename length depending on the OS. Check this answer: https://stackoverflow.com/a/35240286/6352482 , the second approach can be taken here.
Ref: https://en.wikipedia.org/wiki/Filename (contains list of reserved keywords, filename length restrictions for different os)